### ISSUE TITLE
Switch to aiohttp for outgoing HTTP requests

### ISSUE THREAD

**Issue by yuvipanda** [jupyterhub/jupyterhub#2038]
We currently use Tornado's HTTP Client (curl or simple). However, now that we're running on asyncio, we can use the more widely used aiohttp library. Minimum, it gives us a performance boost on PyPy. I also sortof like its interface better.

**Comment by minrk** [comment]
Interesting! I'm skeptical of it being more widely used, but happy to pick it up if it works better. Anecdotally, I've personally found it to be more error-prone than the tornado clients when dealing with large numbers of concurrent requests.

This should be a pretty easy refactor to try out, since we only use the http client directly in a couple of places.

**Comment by willingc** [comment]
Just an FYI, several Python core devs also maintain aiohttp as well as asyncio. It may be worth trying it out especially if the testing tools around it are more robust than Tornado.

**Comment by consideRatio** [comment]
Hi @yuvipanda @minrk @willingc, what are you currently thinking about this? Is aiohttp for outgoing HTTP requests sensible to try out for performance reasons instead of using Tornado's HTTP client to make outbound HTTP requests?

I was issue triaging in jupyterhub/oauthenticator, and there is an issue in there (https://github.com/jupyterhub/oauthenticator/issues/202) as well about this kind of switch. I don't much about this myself.

Hi @willingc :heart: :sunflower:!

**Comment by yuvipanda** [comment]
Yeah, I still think moving to aiohttp is worth a go. Definitely interested in what it gives us re PyPy as well - when I ran perf tests, calling out to libcurl as we do now was a big reason we didn't get a lot of perf gains. It also requires a compiled library, so would be nice to not need it.

**Comment by minrk** [comment]
I'm actually less sure, as I've seen lots *more* timeouts and concurrency issues when I try out aiohttp in other projects than tornado+pycurl, for instance. May still be a fine thing to do, but I think it sould be a pretty low priority (as evidenced by still not tackling it after some years!). I also think tornado's clients are more pleasant to use, but that's a minor hurdle to wrap the nested context managers of aiohttp into a common `utils.fetch()` that we can use instead. So I don't think there is a problem that we have right now that would be solved by switching to aiohttp, and I'm skeptical that performance would be improved.

It's been a while, but I *think* aiodns is required for aiohttp to not just fail all the time under load, so make sure to add `aiohttp[speedups]` as a dependency, not just `aiohttp` if someone is going to explore this. And it would probably be interesting to have some benchmarks for when/whether aiohttp actually improves things in our context.
